src/Http/Routing/src/Internal/MapActionExpressionTreeBuilder.cs

```diff
@@ -0,0 +1,442 @@
+                                           methodCall,
+                                           TargetArg,
+                                           HttpContextParameter);
+                    }
+                    else
+                    {
+                        // ExecuteTask<T>(action(..), httpContext);
+                        body = Expression.Call(
+                                       ExecuteValueTaskOfTMethodInfo.MakeGenericMethod(typeArg),
+                                       methodCall,
+                                       TargetArg,
+                                       HttpContextParameter);
+                    }
+                }
+                else
+                {
+                    // TODO: Handle custom awaitables
+                    throw new NotSupportedException($"Unsupported return type: {method.ReturnType}");
+                }
+            }
+            else if (typeof(IResult).IsAssignableFrom(method.ReturnType))
+            {
+                body = Expression.Call(methodCall, ResultWriteResponseAsync, HttpContextParameter);
+            }
+            else if (method.ReturnType == typeof(string))
+            {
+                body = Expression.Call(StringResultWriteResponseAsync, HttpResponseExpr, methodCall, Expression.Constant(CancellationToken.None));
+            }
+            else
+            {
+                body = Expression.Call(JsonResultWriteResponseAsync, HttpResponseExpr, methodCall, Expression.Constant(CancellationToken.None));
+            }
+
+            Func<object?, HttpContext, Task>? requestDelegate = null;
+
+            if (consumeBodyDirectly)
+            {
+                // We need to generate the code for reading from the body before calling into the delegate
+                var lambda = Expression.Lambda<Func<object?, HttpContext, object?, Task>>(body, TargetArg, HttpContextParameter, DeserializedBodyArg);
+                var invoker = lambda.Compile();
+                object? defaultBodyValue = null;
+
+                if (allowEmptyBody && bodyType!.IsValueType)
+                {
+                    defaultBodyValue = Activator.CreateInstance(bodyType);
+                }
+
+                requestDelegate = async (target, httpContext) =>
+                {
+                    object? bodyValue;
+
+                    if (allowEmptyBody && httpContext.Request.ContentLength == 0)
+                    {
+                        bodyValue = defaultBodyValue;
+                    }
+                    else
+                    {
+                        try
+                        {
+                            bodyValue = await httpContext.Request.ReadFromJsonAsync(bodyType!);
+                        }
+                        catch (IOException ex)
+                        {
+                            LogRequestBodyIOException(httpContext, ex);
+                            httpContext.Abort();
```

Comment:
@pranavkm is this how MVC handles it? I thought it set a 400 response status instead?

Reply:
Yea, why is it aborting the connection?

Reply:
JSON parsing errors result in a 400 response that includes details from the parser. Other IO errors result in 400s. 

IMO, I assumed the entirety of this type was WIP so I hadn't looked at it in detail. We should factor it because this isn't a sustainable format to maintain.

Reply:
This is **very** much a work in progress. I imagine we'll want to make logic like this reusable from different kinds of code-gen and move it out of this class at some point.

As for the status code, it now sets a 400 for InvalidDataExceptions. For aborted request, the status code doesn't matter much since the response doesn't get written to a socket and Kestrel will [set the status code to 0](https://github.com/dotnet/aspnetcore/blob/main/src/Servers/Kestrel/Core/src/Internal/Http/HttpProtocol.cs#L732) before hosting logs it anyway.

Reply:
Hmm, we missed the invalid Json case here. That should also return a 400.